### PR TITLE
day 2 solution

### PR DESCRIPTION
**Pull request by mrsimonfletcher**

### CODE REVIEW

File: years/2020/02/solutions.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  return matches
+}
+
+export const part2Matches = (input: string[]) => {
+  const matches = input.filter((policy: string) => {
+    const [position1, position2, password, stringToMatch] = regex_match(policy)
+
+    if (
+      (password.split('')[parseInt(position1) - 1] == stringToMatch &&
```
Comment (mrseanbaines) [inline years/2020/02/solutions.ts:25]: You don't need to `.split` the string to access individual characters. You could use bracket notation like you would on an array (e.g. `password[0]`) or use `.charAt` (e.g. `password.charAt(0)`) - I tend to prefer the latter, as it's clearer to see that you're working with a string, and not an array

File: years/2020/02/solutions.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        password.split('')[parseInt(position2) - 1] != stringToMatch) ||
+      (password.split('')[parseInt(position1) - 1] != stringToMatch &&
+        password.split('')[parseInt(position2) - 1] == stringToMatch)
```
Comment (mrseanbaines) [inline years/2020/02/solutions.ts:28]: I would defo try to extract some of the repetition here, e.g. create a var above such as `const p1 = parseInt(position1) - 1` and same for `p2`, or even `const p1 = password[Number(position1) - 1]`. Altogether could look something like:

```ts
const p1 = password.charAt(Number(position1) - 1)
const p2 = password.charAt(Number(position2) - 1)

if ((p1 === stringToMatch && p2 !== stringToMatch) || (p1 !== stringToMatch && p2 === stringToMatch)) {
  return true
}
```

File: years/2020/02/solutions.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    return false
+  })
+
+  return matches
+}
+
+function regex_match(policy: string) {
+  const regex = policy.match(/(?<position1>\d*)-(?<position2>\d*)\s(?<string_to_match>[a-zA-Z]):\s(?<password>.*)/)
```
Comment (mrseanbaines) [inline years/2020/02/solutions.ts:40]: Holy moly

Comment (mrseanbaines) [reply years/2020/02/solutions.ts:40]: I don't think I knew you could set variable names for capture groups.. cool!

File: years/2020/02/solutions.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  if (!regex || !regex.groups) throw new Error(`Sorry, couldn't find a match`)
+
+  return [regex.groups.position1, regex.groups.position2, regex.groups.password, regex.groups.string_to_match]
```
Comment (mrseanbaines) [inline years/2020/02/solutions.ts:43]: I think you don't need to access `.groups` and can make this function a bit shorter:

```ts
const [match, position1, position2, stringToMatch, password] = '1-3 a: abcde'.match(/(?\d*)-(?\d*)\s(?[a-zA-Z]):\s(?.*)/)
```
..can probably remove the var names in the regex then and make it even shorter.

The vars in the regex _do_ make it quite clear what's what though, which I like. I think I just don't like the repetition of `regex.groups.*`.. could keep your regex as-is as just destructure `regex.groups`:

```ts
const { position1, position2, password, string_to_match } = regex.groups

return [position1, position2, password, string_to_match]
```

...camelCased though, of course

**Review by mrseanbaines — APPROVED** [review]
:shipit: 

Some of the same comments for this as in day 1 (`let` vs. `const`, blockless `if`s, arrow functions etc.) but omitted from the review comments

File: years/2020/02/solutions.ts (hunk 1 above)
Comment (mrsimonfletcher) [reply years/2020/02/solutions.ts:25]: TIL about `charAt` thanks!

**Pull request merged**